streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -142,148 +144,129 @@ trait ZSink[-R, +E, +A0, -A, +B] { self =>
+  final def as[C](c: => C): ZSink[R, E, A, C] = self.map(_ => c)
 
   /**
    * Replaces any error produced by this sink.
    */
-  final def asError[E1](e1: E1): ZSink[R, E1, A0, A, B] = self.mapError(_ => e1)
+  final def asError[E1](e1: E1): ZSink[R, E1, A, B] = self.mapError(_ => e1)
 
-  private[ZSink] final def chunked[A00 >: A0, A1 <: A](implicit ev: A1 =:= A00): ZSink[R, E, A00, Chunk[A1], B] =
-    new ZSink[R, E, A00, Chunk[A1], B] {
-      type State = (self.State, Chunk[A00])
+  final def chunked: ZSink[R, E, Chunk[A], B] =
+    new ZSink[R, E, Chunk[A], B] {
+      type State = (self.State, Chunk[A])
       val initial = self.initial.map((_, Chunk.empty))
-      def step(state: State, a: Chunk[A1]) =
-        self.stepChunkSlice(state._1, a).map { case (s, chunk) => (s, chunk.map(ev)) }
-      def extract(state: State) = self.extract(state._1).map { case (b, leftover) => (b, leftover ++ state._2) }
+      def step(state: State, a: Chunk[A]) =
+        self.stepChunk(state._1, a).map { case (s, chunk) => (s, chunk) }
+      def extract(state: State) = self.extract(state._1).map { case (b, leftover) => (b, Chunk(leftover, state._2)) }
       def cont(state: State)    = self.cont(state._1)
     }
 
-  private[ZSink] final def collectAll[A00 >: A0, A1 <: A](
-    implicit ev: A00 =:= A1,
-    ev2: A1 =:= A00
-  ): ZSink[R, E, A00, A1, List[B]] =
-    collectAllWith[List[B], A00, A1](List.empty[B])((bs, b) => b :: bs).map(_.reverse)
-
-  private[ZSink] final def collectAllN[A00 >: A0, A1 <: A](
-    i: Int
-  )(implicit ev: A00 =:= A1, ev2: A1 =:= A00): ZSink[R, E, A00, A1, List[B]] =
-    collectAllWith[(List[B], Int), A00, A1]((Nil, 0)) {
-      case ((bs, len), b) =>
-        (b :: bs, len + 1)
-    }.untilOutput {
-      case (_, len) =>
-        len >= i
-    }.map(_.getOrElse((Nil, 0))._1.reverse)
-
-  private[ZSink] final def collectAllWith[S, A00 >: A0, A1 <: A](
-    z: S
-  )(f: (S, B) => S)(implicit ev: A00 =:= A1, ev2: A1 =:= A00): ZSink[R, E, A00, A1, S] =
-    collectAllWhileWith[S, A00, A1](_ => true)(z)(f)
-
-  private[ZSink] final def collectAllWhile[A00 >: A0, A1 <: A](
-    p: A1 => Boolean
-  )(implicit ev: A00 =:= A1, ev2: A1 =:= A00): ZSink[R, E, A00, A1, List[B]] =
-    collectAllWhileWith[List[B], A00, A1](p)(List.empty[B])((bs, b) => b :: bs)
-      .map(_.reverse)
+  final def collectAll: ZSink[R, E, A, List[B]] =
+    collectAllWith[List[B]](List.empty[B])((bs, b) => b :: bs).map(_.reverse)
 
-  private[ZSink] final def collectAllWhileWith[S, A00 >: A0, A1 <: A](
-    p: A1 => Boolean
-  )(z: S)(f: (S, B) => S)(implicit ev: A00 =:= A1, ev2: A1 =:= A00): ZSink[R, E, A00, A1, S] =
-    new ZSink[R, E, A00, A1, S] {
-      type State = (S, self.State, Boolean, Chunk[A00])
+  final def collectAllN(i: Int): ZSink[R, E, A, List[B]] =
+    new ZSink[R, E, A, List[B]] {
+      case class State(s: self.State, bs: List[B], n: Int, leftover: Chunk[A], dirty: Boolean)
```

Comment:
This is totally new to me. One could implement abstract types as classes. Seems so simple and logical but wow. :boom: